### ISSUE TITLE
Gradient checkpointing support in C++ API

### ISSUE THREAD

**Issue by Elysion-tcfa** [pytorch/pytorch#52744]
## 🚀 Feature
Gradient checkpointing feature is available as torch.utils.checkpoint.* in Python API. However, after an exhaustive search, I can't find a C++ equivalent of this feature. Self-implementing also seems difficult because C++ Node class is quite different from Python torch.autograd.Function class.

## Motivation

Due to a subtle reason I have to use C++ for model training. Everything works fine in using PyTorch C++ APIs since they mostly resembles their Python counterparts. However, one thing which is essential but missing is the gradient checkpointing feature.

## Pitch

The gradient checkpointing feature has an equivalent in C++: its functions most desirably reside in torch::utils::checkpoint namespace and resemble their Python counterparts `checkpoint` and `checkpoint_sequential`.

## Alternatives

Manually checkpointing the gradients is possible at least in my case, by replicating the logic in CheckpointFunction. However, since I can't hack the backward pass, I have to manually do backward() many times from each checkpointed output.

## Additional context

None
